In [2]:
%matplotlib notebook
import pandas as pd
import numpy as np
import seaborn as sn
import matplotlib.pyplot as plt

# df (my data frame)
df = pd.read_csv("Z:/MachLearning/base_datos_2008.csv")
from sklearn.model_selection import train_test_split


## Exploración y depuración 

In [13]:
df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [3]:
print("Cantidad vacios de WeatherDelay ",len(df[pd.isna(df["WeatherDelay"])]))

print("Cantidad vacios de LateAircraftDelay ",len(df[pd.isna(df["LateAircraftDelay"])]))

print("Cantidad vacios de Diverted", len(df[pd.isna(df["Diverted"])]))

print("Cantidad vacios de CarrierDelay",len(df[pd.isna(df["CarrierDelay"])]))

print("Cantidad vacios de ArrDelay",len(df[pd.isna(df["ArrDelay"])]))

print("Cantidad vacios de DayOfWeek",len(df[pd.isna(df["DayOfWeek"])]))

print("Cantidad vacios de TaxiIn",len(df[pd.isna(df["TaxiIn"])]))

df = df[df["WeatherDelay"].notna() &
        df["LateAircraftDelay"].notna() &
        df["Diverted"].notna() &
        df["CarrierDelay"].notna() &
        df["ArrDelay"].notna() &
        df["DayofMonth"].notna() &
        df["DayOfWeek"].notna()]


Cantidad vacios de WeatherDelay  5484993
Cantidad vacios de LateAircraftDelay  5484993
Cantidad vacios de Diverted 0
Cantidad vacios de CarrierDelay 5484993
Cantidad vacios de ArrDelay 154699
Cantidad vacios de DayOfWeek 0
Cantidad vacios de TaxiIn 151649


In [4]:
feature_names_flights = ["Distance","Diverted","CarrierDelay","WeatherDelay","LateAircraftDelay"]
X_flights = df[feature_names_flights]
y_flights = df['ArrDelay']

## KNeighborsRegressor

**KNeighborsRegressor:** **Para este modelo acotamos el tamaño del dataset por motivos prácticos ya que este modelo su entrenamiento es
costoso para un dataset muy grande como en este caso con 1,5 M de filas.**

In [8]:

from sklearn.neighbors import KNeighborsRegressor
X_train, X_test, y_train, y_test = train_test_split(X_flights[0:5000], y_flights[0:5000], random_state = 10)

knnreg = KNeighborsRegressor(n_neighbors = 10).fit(X_train, y_train)

print('test score: {:.3f}'
     .format(knnreg.score(X_test, y_test)))

test score: 0.762


In [16]:
from sklearn.neighbors import KNeighborsRegressor

X_train, X_test, y_train, y_test = train_test_split(X_flights[0:5000], y_flights[0:5000], random_state = 10)

for  K in [1,2,3,5,6,6,7,10,20]:
    knnreg = KNeighborsRegressor(n_neighbors = K).fit(X_train, y_train)
    print("{} KnnReg\n train score: {:.2f} test score: {:.2f}".format(K,    
              knnreg.score(X_train, y_train),knnreg.score(X_test, y_test)))


1 KnnReg
 train score: 0.97 test score: 0.71
2 KnnReg
 train score: 0.93 test score: 0.79
3 KnnReg
 train score: 0.90 test score: 0.80
5 KnnReg
 train score: 0.86 test score: 0.80
6 KnnReg
 train score: 0.85 test score: 0.79
6 KnnReg
 train score: 0.85 test score: 0.79
7 KnnReg
 train score: 0.83 test score: 0.78
10 KnnReg
 train score: 0.80 test score: 0.76
20 KnnReg
 train score: 0.71 test score: 0.70


***Podemos ver que en esta situacion el mejor score alcanzado es con una cantidad de vecinos igual a 5***

## LinearRegression

In [9]:
from sklearn.linear_model import LinearRegression

X_train, X_test, y_train, y_test = train_test_split(X_flights, y_flights,
                                                   random_state = 10)
linreg = LinearRegression().fit(X_train, y_train)

print('Regresion lineal (ridge) :valor de intercept calculado: {}'
     .format(linreg.intercept_))
print('Regresion lineal (ridge) :Coeficientes calculados:\n{}'
     .format(linreg.coef_))
print('R-squared score (training): {:.3f}'
     .format(linreg.score(X_train, y_train)))
print('R-squared score (test): {:.3f}'
     .format(linreg.score(X_test, y_test)))

Regresion lineal (ridge) :valor de intercept calculado: 21.31939699033586
Regresion lineal (ridge) :Coeficientes calculados:
[0.00128099 0.         0.86847563 0.95413467 0.86289565]
R-squared score (training): 0.701
R-squared score (test): 0.710


## Ridge Regression

In [10]:

from sklearn.linear_model import Ridge
X_train, X_test, y_train, y_test = train_test_split(X_flights, y_flights, random_state = 10)

linridge = Ridge(alpha=30.0).fit(X_train, y_train)
print('Regresion lineal (ridge) :valor de intercept calculado: {}'
     .format(linridge.intercept_))
print('Regresion lineal (ridge) :Coeficientes calculados:\n{}'
     .format(linridge.coef_))
print('R-squared score (training): {:.3f}'
     .format(linridge.score(X_train, y_train)))
print('R-squared score (test): {:.3f}'
     .format(linridge.score(X_test, y_test)))


Regresion lineal (ridge) :valor de intercept calculado: 21.319397952524795
Regresion lineal (ridge) :Coeficientes calculados:
[0.00128099 0.         0.86847561 0.9541346  0.86289564]
R-squared score (training): 0.701
R-squared score (test): 0.710


In [15]:
print('Variación del hyperparametro alpha  \n')
for this_alpha in [0, 1, 10, 20, 50, 100]:
    linridge = Ridge(alpha = this_alpha).fit(X_train, y_train)
    r2_train = linridge.score(X_train, y_train)
    r2_test = linridge.score(X_test, y_test)
    num_coeff_bigger = np.sum(abs(linridge.coef_) > 1.0)
    print('Alpha = {:.2f}\nnum abs(coeff) > 1.0: {}, \
r-squared training: {:.2f}, r-squared test: {:.2f}\n'
         .format(this_alpha, num_coeff_bigger, r2_train, r2_test))
    
    

Variación del hyperparametro alpha  

Alpha = 0.00
num abs(coeff) > 1.0: 2, r-squared training: 0.67, r-squared test: 0.68

Alpha = 1.00
num abs(coeff) > 1.0: 0, r-squared training: 0.70, r-squared test: 0.71

Alpha = 10.00
num abs(coeff) > 1.0: 0, r-squared training: 0.70, r-squared test: 0.71

Alpha = 20.00
num abs(coeff) > 1.0: 0, r-squared training: 0.70, r-squared test: 0.71

Alpha = 50.00
num abs(coeff) > 1.0: 0, r-squared training: 0.70, r-squared test: 0.71

Alpha = 100.00
num abs(coeff) > 1.0: 0, r-squared training: 0.70, r-squared test: 0.71



***Vemos que al aumentar alpha se alcanza a un score fijo lo cual es dudoso este comportamiento...***

## Cross-validation

***Para la evaluación de desempeño de nuestro modelo KnnReg usamos validacion cruzada que consta de distintas diviciones de nuestro
dataset, en este caso usamos tres diviciones.***

In [14]:

from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsRegressor

knnreg = KNeighborsRegressor(n_neighbors = 5)
X = X_flights[0:5000]
y = y_flights[0:5000]
cv_scores = cross_val_score(knnreg, X, y)

print('Cross-validation scores (3-fold):', cv_scores)
print('Promedio cross-validation score (3-fold): {:.3f}'
     .format(np.mean(cv_scores)))

Cross-validation scores (3-fold): [0.7650618  0.81212115 0.81515475 0.80637413 0.69463459]
Mean cross-validation score (3-fold): 0.779


***Lo cual nos dice que en uno de los peores comportamiento de nuestro modelo seria 0.6946 y 0.812 para el mejor comportamiento. Con un promedio de 0.779 para estas tres particiones***